File: server/ruoyi-modules/jshpos-promotion/src/test/java/com/jingshanghui/pos/promotion/domain/PromotionGoldenVectorTest.java
package com.jingshanghui.pos.promotion.domain;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.jingshanghui.pos.promotion.domain.PromotionModels.*;
import org.junit.jupiter.api.Test;

import java.math.BigDecimal;
import java.nio.file.Files;
import java.nio.file.Path;
import java.time.OffsetDateTime;
import java.util.*;

import static org.assertj.core.api.Assertions.assertThat;

/** 共享黄金向量是服务端与 Flutter 跨端一致性的唯一输入。 */
class PromotionGoldenVectorTest {
    private final ObjectMapper objectMapper = new ObjectMapper();
    private final PromotionEngine engine = new PromotionEngine();

    @Test
    void shouldMatchEveryPrm1GoldenVectorExactly() throws Exception {
        Path path = Path.of(System.getProperty("user.dir")).toAbsolutePath();
        while (path != null && !Files.isDirectory(path.resolve("contracts"))) path = path.getParent();
        path = path == null ? Path.of("missing") : path.resolve(Path.of("contracts", "t2", "gate5a",
            "test-vectors", "promotion-golden-vectors-v1.json"));
        assertThat(Files.isRegularFile(path)).isTrue();
        JsonNode root = objectMapper.readTree(path.toFile());
        assertThat(root.path("engineVersion").asText()).isEqualTo(PromotionEngine.ENGINE_VERSION);
        assertThat(root.path("scenarios")).hasSize(17);
        for (JsonNode scenario : root.path("scenarios")) {
            QuoteResult actual = engine.quote(toRequest(scenario));
            JsonNode expected = scenario.path("expected");
            assertThat(actual.grossAmountMinor()).as(scenario.path("id").asText())
                .isEqualTo(expected.path("grossAmountMinor").asLong());
            assertThat(actual.discountAmountMinor()).isEqualTo(expected.path("discountAmountMinor").asLong());
            assertThat(actual.payableAmountMinor()).isEqualTo(expected.path("payableAmountMinor").asLong());
            Map<String, Long> expectedLines = new LinkedHashMap<>();
            expected.path("lineDiscounts").fields().forEachRemaining(entry ->
                expectedLines.put(entry.getKey(), entry.getValue().asLong()));
            assertThat(actual.lineDiscounts()).containsExactlyEntriesOf(expectedLines);
            assertThat(actual.appliedRuleIds()).containsExactlyElementsOf(strings(expected.path("appliedRuleIds")));
            List<Explanation> expectedExplanations = new ArrayList<>();
            expected.path("explanations").forEach(node -> expectedExplanations.add(new Explanation(
                node.path("sourceId").asText(), node.path("code").asText())));
            assertThat(actual.explanations()).containsExactlyElementsOf(expectedExplanations);
            assertThat(actual.grossAmountMinor() - actual.discountAmountMinor())
                .isEqualTo(actual.payableAmountMinor());
        }
    }

    private QuoteRequest toRequest(JsonNode scenario) {
        List<BasketLine> lines = new ArrayList<>();
        scenario.path("lines").forEach(node -> lines.add(new BasketLine(node.path("lineId").asText(),
            node.path("lineNo").asInt(), node.path("skuId").asLong(), nullableLong(node, "categoryId"),
            nullableLong(node, "brandId"), new BigDecimal(node.path("quantity").asText()),
            node.path("unitPriceMinor").asLong())));
        List<RuleVersion> rules = new ArrayList<>();
        scenario.path("rules").forEach(node -> rules.add(new RuleVersion(node.path("ruleVersionId").asText(),
            RuleType.valueOf(node.path("ruleType").asText()), node.path("priority").asInt(),
            StackMode.valueOf(node.path("stackMode").asText()), nullableText(node, "exclusiveGroup"),
            OffsetDateTime.parse(node.path("effectiveFrom").asText()),
            nullableDateTime(node, "effectiveTo"), scope(node.path("scope")), benefit(node.path("benefit")))));
        return new QuoteRequest(OffsetDateTime.parse(scenario.path("businessTime").asText()),
            scenario.path("storeId").asLong(), scenario.path("channel").asText(), lines, rules);
    }

    private RuleScope scope(JsonNode node) {
        return new RuleScope(longs(node.path("skuIds")), longs(node.path("categoryIds")),
            longs(node.path("brandIds")), longs(node.path("storeIds")),
            Set.copyOf(strings(node.path("channels"))), integers(node.path("businessDays")));
    }

    private RuleBenefit benefit(JsonNode node) {
        List<BundleComponent> components = new ArrayList<>();
        node.path("bundleComponents").forEach(value -> components.add(new BundleComponent(
            value.path("skuId").asLong(), new BigDecimal(value.path("quantity").asText()))));
        return new RuleBenefit(nullableLong(node, "amountMinor"), decimal(node, "discountRate"),
            node.hasNonNull("nth") ? node.path("nth").asInt() : null, nullableLong(node, "thresholdMinor"),
            decimal(node, "thresholdQuantity"), nullableLong(node, "bundlePriceMinor"), components);
    }

    private Set<Long> longs(JsonNode node) {
        Set<Long> values = new LinkedHashSet<>();
        node.forEach(value -> values.add(value.asLong()));
        return values;
    }

    private Set<Integer> integers(JsonNode node) {
        Set<Integer> values = new LinkedHashSet<>();
        node.forEach(value -> values.add(value.asInt()));
        return values;
    }

    private List<String> strings(JsonNode node) {
        List<String> values = new ArrayList<>();
        node.forEach(value -> values.add(value.asText()));
        return values;
    }

    private Long nullableLong(JsonNode node, String field) {
        return node.hasNonNull(field) ? node.path(field).asLong() : null;
    }

    private String nullableText(JsonNode node, String field) {
        return node.hasNonNull(field) ? node.path(field).asText() : null;
    }

    private BigDecimal decimal(JsonNode node, String field) {
        return node.hasNonNull(field) ? new BigDecimal(node.path(field).asText()) : null;
    }

    private OffsetDateTime nullableDateTime(JsonNode node, String field) {
        return node.hasNonNull(field) ? OffsetDateTime.parse(node.path(field).asText()) : null;
    }
}

File: pos-flutter/test/gate5a/promotion_golden_vector_test.dart
import 'dart:convert';
import 'dart:io';

import 'package:flutter_test/flutter_test.dart';
import 'package:jshpos_pos/features/promotion/domain/promotion_engine.dart';

void main() {
  test('Flutter matches every shared PRM-001 golden vector exactly', () {
    final file = File(
      '../contracts/t2/gate5a/test-vectors/promotion-golden-vectors-v1.json',
    );
    expect(file.existsSync(), isTrue);
    final root = jsonDecode(file.readAsStringSync()) as Map<String, Object?>;
    expect(root['engineVersion'], PromotionEngine.engineVersion);
    final scenarios = root['scenarios']! as List<Object?>;
    expect(scenarios, hasLength(17));
    final engine = PromotionEngine();
    for (final raw in scenarios) {
      final scenario = raw! as Map<String, Object?>;
      final result = engine.quote(
        businessTime: DateTime.parse(scenario['businessTime']! as String),
        storeId: scenario['storeId']! as String,
        channel: scenario['channel']! as String,
        lines: (scenario['lines']! as List<Object?>).map(_line).toList(),
        rules: (scenario['rules']! as List<Object?>).map(_rule).toList(),
      );
      final expected = scenario['expected']! as Map<String, Object?>;
      expect(
        result.grossAmountMinor,
        expected['grossAmountMinor'],
        reason: scenario['id']! as String,
      );
      expect(
        result.discountAmountMinor,
        expected['discountAmountMinor'],
        reason: scenario['id']! as String,
      );
      expect(
        result.payableAmountMinor,
        expected['payableAmountMinor'],
        reason: scenario['id']! as String,
      );
      expect(
        result.lineDiscounts,
        (expected['lineDiscounts']! as Map<String, Object?>).map(
          (key, value) => MapEntry(key, value! as int),
        ),
      );
      expect(
        result.appliedRuleIds,
        (expected['appliedRuleIds']! as List<Object?>).cast<String>(),
      );
      expect(
        result.explanations,
        (expected['explanations']! as List<Object?>).map((item) {
          final value = item! as Map<String, Object?>;
          return PromotionExplanation(
            value['sourceId']! as String,
            value['code']! as String,
          );
        }).toList(),
      );
      expect(
        result.adjustments.fold(0, (sum, item) => sum + item.amountMinor),
        result.discountAmountMinor,
      );
    }
  });
}

PromotionLine _line(Object? raw) {
  final value = raw! as Map<String, Object?>;
  return PromotionLine(
    lineId: value['lineId']! as String,
    lineNo: value['lineNo']! as int,
    skuId: value['skuId']! as String,
    categoryId: value['categoryId'] as String?,
    brandId: value['brandId'] as String?,
    quantity: ExactDecimal.parse(value['quantity']! as String, maximumScale: 6),
    unitPriceMinor: value['unitPriceMinor']! as int,
  );
}

PromotionRule _rule(Object? raw) {
  final value = raw! as Map<String, Object?>;
  final scope = value['scope']! as Map<String, Object?>;
  final benefit = value['benefit']! as Map<String, Object?>;
  return PromotionRule(
    ruleVersionId: value['ruleVersionId']! as String,
    ruleType: _ruleType(value['ruleType']! as String),
    priority: value['priority']! as int,
    stackMode: _stackMode(value['stackMode']! as String),
    exclusiveGroup: value['exclusiveGroup'] as String?,
    effectiveFrom: DateTime.parse(value['effectiveFrom']! as String),
    effectiveTo: DateTime.parse(value['effectiveTo']! as String),
    scope: PromotionScope(
      skuIds: _strings(scope['skuIds']),
      categoryIds: _strings(scope['categoryIds']),
      brandIds: _strings(scope['brandIds']),
      storeIds: _strings(scope['storeIds']),
      channels: _strings(scope['channels']),
      businessDays: _integers(scope['businessDays']),
    ),
    benefit: PromotionBenefit(
      amountMinor: benefit['amountMinor'] as int?,
      discountRate: _decimal(benefit['discountRate'], 8),
      nth: benefit['nth'] as int?,
      thresholdMinor: benefit['thresholdMinor'] as int?,
      thresholdQuantity: _decimal(benefit['thresholdQuantity'], 6),
      bundlePriceMinor: benefit['bundlePriceMinor'] as int?,
      bundleComponents:
          ((benefit['bundleComponents'] as List<Object?>?) ?? const []).map((
            item,
          ) {
            final component = item! as Map<String, Object?>;
            return BundleComponent(
              component['skuId']! as String,
              ExactDecimal.parse(
                component['quantity']! as String,
                maximumScale: 6,
              ),
            );
          }).toList(),
    ),
  );
}

Set<String> _strings(Object? value) =>
    ((value as List<Object?>?) ?? const []).cast<String>().toSet();

Set<int> _integers(Object? value) =>
    ((value as List<Object?>?) ?? const []).cast<int>().toSet();
ExactDecimal? _decimal(Object? value, int scale) => value == null
    ? null
    : ExactDecimal.parse(value as String, maximumScale: scale);

PromotionRuleType _ruleType(String value) => switch (value) {
  'SPECIAL_PRICE' => PromotionRuleType.specialPrice,
  'PERCENT_OFF' => PromotionRuleType.percentOff,
  'AMOUNT_OFF' => PromotionRuleType.amountOff,
  'NTH_ITEM_DISCOUNT' => PromotionRuleType.nthItemDiscount,
  'BUNDLE_PRICE' => PromotionRuleType.bundlePrice,
  'THRESHOLD_AMOUNT_OFF' => PromotionRuleType.thresholdAmountOff,
  'THRESHOLD_QUANTITY_OFF' => PromotionRuleType.thresholdQuantityOff,
  _ => throw FormatException('unknown rule $value'),
};

PromotionStackMode _stackMode(String value) => switch (value) {
  'EXCLUSIVE' => PromotionStackMode.exclusive,
  'STACKABLE' => PromotionStackMode.stackable,
  'BEST_OF_GROUP' => PromotionStackMode.bestOfGroup,
  _ => throw FormatException('unknown stack mode $value'),
};
